src/Lookup/DispatchingEntityLookup.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Lookup;
+
+use Wikibase\DataModel\Entity\EntityDocument;
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\EntityIdParser;
+use Wikimedia\Assert\Assert;
+use Wikimedia\Assert\ParameterAssertionException;
+
+/**
+ * Delegates lookup to the repository-specific EntityLookup
+ * based on the name of the repository EntityId is coming from.
+ *
+ * @since 3.7
+ *
+ * @license GPL-2.0+
+ */
+class DispatchingEntityLookup implements EntityLookup {
+
+	/**
+	 * @var EntityLookup[]
+	 */
+	private $lookups;
+
+	/**
+	 * @var EntityIdParser
+	 */
+	private $parser;
+
+	/**
+	 * @param EntityLookup[] $lookups associative array with repository names (strings) as keys
+	 *                                and EntityLookup objects as values. Empty-string key
+	 *                                defines lookup for the local repository.
+	 *
+	 * @throws ParameterAssertionException
+	 */
+	public function __construct( array $lookups, EntityIdParser $parser ) {
+		Assert::parameter(
+			!empty( $lookups ) && array_key_exists( '', $lookups ),
+			'$lookups',
+			'must must not be empty and must contain an empty-string key'
+		);
+		Assert::parameterElementType( EntityLookup::class, $lookups, '$lookups' );
+		Assert::parameterElementType( 'string', array_keys( $lookups ), 'array_keys( $lookups )' );
+		foreach ( array_keys( $lookups ) as $repositoryName ) {
```

Comment:
I've submitted something for this as https://github.com/wmde/WikibaseDataModelServices/pull/155.
To avoid creating yet more mess with all commits in this PR, I'd rather wait for that PR to at least be in shape ready to be merged and then do changes here (and in other places). 

Please do not "approve the change" (do not remove -1 in github's speak) before it is done so we don't forget about this.